joomla/joomla-cms
# Multiple showon

### Description

This PR adds the ability to use a multiple field dependencies in showon form field attribute.
`showon="field1:value1[OPERATOR]field2:value1,value2"`

This is needed, for instance, in https://github.com/joomla/joomla-cms/pull/8442 , but could really be used in many cases.
### How to test

Apply the patch and refresh javascript cache.
1. Test if all current showon fields (config, recaptcha plugin, tinymce, etc) are working properly
   Hint: To check where showon is being used use (in Linux): `egrep -R 'showon=' /path/to/joomla/`
2. Change a field in `/administrator/components/com_config/model/form/application.xml` to use multiple dependencies, for instance, replace:
   `showon="mailer:sendmail"`
   for
   `showon="mailonline:1[AND]mailer:sendmail"`
   or
   `showon="mailonline:1[OR]mailer:sendmail"`

And test if `jform[sendmail]` field (Sendmail Path) in global config only appears when `jform[mailonline]` (Send Mail) is set to "Yes" and/or  `jform[mailer]` (Mailer) is set to "Sendmail".

Other field showon test examples:

```
<field name="field1" type="checkbox" label="field1" value="1" default="0" />
<field name="field2" type="radio" label="field2" default="0" class="btn-group btn-group-yesno"><option value="1">1</option><option value="0">0</option></field>
<field name="field3" type="list" label="field3" default="0"><option value="1">1</option><option value="0">0</option></field>
<field name="field4" type="text" label="field4" default="0" />
<field name="field5" type="checkboxes" label="field5" default="0"><option value="2">2</option><option value="1">1</option><option value="0">0</option></field>
<field name="field6" type="list" label="field6" default="0" multiple="true"><option value="0">0</option><option value="1">1</option></field>

<field name="check1" type="text" label="shows on field1=1" showon="field1:1" /> 
<field name="check2" type="text" label="shows on field1=1 AND field1=1" showon="field1:1[AND]field2:1" />               
<field name="check3" type="text" label="shows on field1=1 OR field2=1" showon="field1:1[OR]field2:1" /> 
<field name="check4" type="text" label="shows on field1=1 OR field2=1 OR field3=1" showon="field1:1[OR]field2:1[OR]field3:1" /> 
<field name="check5" type="text" label="shows on field1=1 AND field2=1 AND field3=1" showon="field1:1[AND]field2:1[AND]field3:1" />
<field name="check6" type="text" label="shows on field1=1 AND field2=1 OR field3=1" showon="field1:1[AND]field2:1[OR]field3:1" />   
<field name="check7" type="text" label="shows on field1=1 OR field2=1 AND field3=1" showon="field1:1[OR]field2:1[AND]field3:1" />   
<field name="check8" type="text" label="shows on field1=1 AND field2=1 AND field3=1 AND field4=1" showon="field1:1[AND]field2:1[AND]field3:1[AND]field4:1" />   
<field name="check9" type="text" label="shows on field5 as 0" showon="field5:0" />
<field name="check10" type="text" label="shows on field5 as 1" showon="field5:1" />
<field name="check11" type="text" label="shows on field5 as 2" showon="field5:2" />
<field name="check12" type="text" label="shows on field5 as 1 or 0" showon="field5:1,0" />
<field name="check13" type="text" label="shows on field5 as 2,1 or 0" showon="field5:2,1,0" />
<field name="check14" type="text" label="shows on field1=1 AND field5 as 2 or 1" showon="field1:1[AND]field5:2,1" />
<field name="check15" type="text" label="shows on field6 as 0" showon="field6:0" />
<field name="check16" type="text" label="shows on field6 as 1 or 0" showon="field6:1,0" />
```
### BC

After update, just refresh javascript cache to refresh cms.js file (Ctrl + F5).

I don't know if i used some form classes properly. Please check code.
Improvements are welcome.
@dgt41 @infograf768
